### PR TITLE
feat(server): add OIDC for AFFiNE

### PR DESCRIPTION
**Pull request by IamTaoChen**
## How-to-Use
```
  OIDC_ISSUER=
  OIDC_CLIENT_ID=
  OIDC_CLIENT_SECRET=
  OIDC_SCOPE=openid email profile
  OIDC_CLAIM_MAP_ID=preferred_username
  OIDC_CLAIM_MAP_EMAIL=email
```

## TO-DO:

- [ ] Add OIDC icon at `packages/frontend/core/src/components/affine/auth/oauth.tsx`

## Bug:(solved)
~~I can get the userinfo by OIDC plugin but there is an error I don't know how to deal with~~
```
http://localhost:8080/signIn?redirect_uri=%2Fworkspace%2FJDJv6ZMiwVlrCOZMyqPda%2Fall&error=%0AInvalid+%60prisma.user.create%28%29%60+invocation%3A%0A%0A%0AAn+operation+failed+because+it+depends+on+one+or+more+records+that+were+required+but+not+found.+No+%27Features%27+record%28s%29+%28needed+to+inline+the+relation+on+%27UserFeatures%27+record%28s%29%29+was+found+for+a+nested+connect+on+one-to-many+relation+%27FeaturesToUserFeatures%27.
```
-->
```
http://localhost:8080/signIn?redirect_uri=/workspace/JDJv6ZMiwVlrCOZMyqPda/all&error=
Invalid `prisma.user.create()` invocation:

An operation failed because it depends on one or more records that were required but not found. No 'Features' record(s) (needed to inline the relation on 'UserFeatures' record(s)) was found for a nested connect on one-to-many relation 'FeaturesToUserFeatures'.
```

### CODE REVIEW

**Comment by forehalo** [comment]
you may run `yarn workspace @affine/server data-migration run` before testing

**Comment by IamTaoChen** [comment]
> yarn workspace @affine/server data-migration run

yeah, now it works.

**Review by Brooooooklyn — APPROVED** [review]

File: packages/backend/server/src/plugins/oauth/providers/oidc.ts [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+      })
+      .catch(error => {
+        this.logger.error('Failed to load OIDC configuration:', error);
+      });
+  }
+
+  private async loadOIDCConfigurationAsync(): Promise<void> {
+    const issuer = this.config.issuer;
+    if (!issuer) {
+      this.logger.error('Issuer is not defined in the configuration.');
```
Comment (darkskygit) [inline packages/backend/server/src/plugins/oauth/providers/oidc.ts]: throw error directly

File: packages/backend/server/src/plugins/oauth/providers/oidc.ts [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+      const fullConfig = await response.json();
+      this.oidcConfig = {
+        authorization_endpoint: fullConfig.authorization_endpoint,
+        token_endpoint: fullConfig.token_endpoint,
+        userinfo_endpoint: fullConfig.userinfo_endpoint,
+        end_session_endpoint: fullConfig.end_session_endpoint,
+      };
+    } catch (error) {
+      this.logger.error(
```
Comment (darkskygit) [inline packages/backend/server/src/plugins/oauth/providers/oidc.ts]: throw error directly, we already catch the error in constructor

File: packages/backend/server/src/plugins/oauth/providers/oidc.ts [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+        },
+      });
+
+      if (!response.ok) {
+        throw new Error(
+          `Failed to fetch OIDC configuration: ${response.statusText}`
+        );
+      }
+
+      const fullConfig = await response.json();
```
Comment (darkskygit) [inline packages/backend/server/src/plugins/oauth/providers/oidc.ts]: make a field verify for json based on zod

File: packages/backend/server/src/plugins/oauth/providers/oidc.ts [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+        return {
+          accessToken: tokenResponse.access_token,
+          refreshToken: tokenResponse.refresh_token,
+          expiresAt: new Date(Date.now() + tokenResponse.expires_in * 1000),
+          scope: tokenResponse.scope,
+        };
+      } else {
+        const errorResponse = await response.json();
+        throw new Error(
+          `Server responded with non-success code ${response.status}, ${JSON.stringify(errorResponse)}`
```
Comment (darkskygit) [inline packages/backend/server/src/plugins/oauth/providers/oidc.ts]: too detailed for returned error, i prefer print the full error in console, and return the error category and status to frontend

File: packages/backend/server/src/plugins/oauth/providers/oidc.ts [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+          name: this.config.args?.claim_name || 'name',
+        };
+        const userinfo = this.mapUserInfo(user, claimsMap);
+        return {
+          id: userinfo.id,
+          email: userinfo.email,
+        };
+      } else {
+        const errorText = await response.text();
+        throw new Error(
```
Comment (darkskygit) [inline packages/backend/server/src/plugins/oauth/providers/oidc.ts]: same as above

File: packages/backend/server/src/plugins/oauth/types.ts [hunk 6]
```diff
@@ -1,15 +1,31 @@
...
   Google = 'google',
   GitHub = 'github',
+  OIDC = 'oidc',
 }
 
 export interface OAuthConfig {
   enabled: boolean;
-  providers: Partial<{ [key in OAuthProviderName]: OAuthProviderConfig }>;
+  providers: Partial<
+    { [key in OAuthProviderName]: OAuthProviderConfig } & {
```
Comment (darkskygit) [inline packages/backend/server/src/plugins/oauth/types.ts]: make a type mapping for OAuthProviderName => OAuthProviderConfig / OAuthOIDCProviderConfig ?

**Review by darkskygit — APPROVED** [review]

**Pull request merged**